### PR TITLE
fix(DatePicker) Address datepicker crashing when manually entering year

### LINKED ISSUE #10135: Bug - DatePicker - Datepicker crashes when manually entering an invalid year
DatePicker component does not allow manual entry of year:

https://github.com/patternfly/patternfly-react/assets/11218376/9cbbfd9c-16b1-490f-bd70-c38e9c622fe6

### PR DESCRIPTION
**Pull request by ibolton336**
**What**: Closes #10135

Before:

https://github.com/patternfly/patternfly-react/assets/11218376/5e91bdab-833e-4124-9acd-d8115492e0f7

After:

https://github.com/patternfly/patternfly-react/assets/11218376/3142a649-c3d2-4935-b76d-25c2f5e2d1b0

### CODE REVIEW

File: packages/react-core/src/components/CalendarMonth/CalendarMonth.tsx [hunk 1]
```diff
@@ -217,6 +222,31 @@ export const CalendarMonth = ({
   const changeMonth = (newMonth: number, newYear?: number) =>
     new Date(newYear ?? focusedDate.getFullYear(), newMonth, 1);
 
+  const MIN_YEAR = 1900;
+  const MAX_YEAR = 2100;
+
+  const handleYearInputChange = (event: React.FormEvent<HTMLInputElement>) => {
+    const yearStr = event.currentTarget.value;
+
+    if (/^\d{0,4}$/.test(yearStr)) {
```
Comment (thatblindgeye) [inline packages/react-core/src/components/CalendarMonth/CalendarMonth.tsx:231]: This is more of a nit, but just to reduce some of the nested if blocks here we could instead return early if `!/^\d{0,4}$/.test(yearStr)`.

File: packages/react-core/src/components/CalendarMonth/CalendarMonth.tsx [hunk 2]
```diff
@@ -217,6 +222,31 @@ export const CalendarMonth = ({
...
+      setYearInput(yearStr);
+
+      if (yearStr.length === 4) {
+        const yearNum = Number(yearStr);
+        if (yearNum >= MIN_YEAR && yearNum <= MAX_YEAR) {
+          const newDate = changeYear(yearNum);
+          setFocusedDate(newDate);
+          setHoveredDate(newDate);
+          setShouldFocus(false);
```
Comment (thatblindgeye) [inline packages/react-core/src/components/CalendarMonth/CalendarMonth.tsx:240]: We need to keep the previous blur call in for FireFox. Without it if a day is selected then the year input scrolls arrows are clicked, pressing Tab acts as though focus is still on the day button rather than the year input.

```suggestion
          setShouldFocus(false);
          // We need to unfocus a date when changing year via up/down arrows in FireFox, otherwise focus doesn't go to the year input like expected
          focusRef.current?.blur(); 
```

File: packages/react-core/src/components/CalendarMonth/CalendarMonth.tsx [hunk 3]
```diff
@@ -217,6 +222,34 @@ export const CalendarMonth = ({
...
+      const yearNum = Number(yearStr);
+
+      if (yearNum >= MIN_YEAR && yearNum <= MAX_YEAR) {
+        const newDate = changeYear(yearNum);
+        setFocusedDate(newDate);
+        setHoveredDate(newDate);
+        setShouldFocus(false);
+
+        // We need to unfocus a date when changing year via up/down arrows in FireFox, otherwise focus doesn't go to the year input like expected
+        focusRef.current?.blur();
```
Comment (thatblindgeye) [inline packages/react-core/src/components/CalendarMonth/CalendarMonth.tsx:245]: Could've sworn in testing yesterday that this worked as originally intended. Noticing in looking again this morning that firing this blur when clicking the scroll buttons in FF has some new(?) odd behavior:

https://github.com/patternfly/patternfly-react/assets/70952936/101078cf-cf6f-4c56-a5e1-5965262f6fc1

Maybe I had just happened to test when the "next month" button received focus correctly.

TLDR is instead we could replace this blur call on the focusRef, and instead `focus()` the TextInput:

```suggestion
        // We need to manually focus the year input in FireFox when the scroll buttons are clicked, as FireFox doesn't place focus automatically
        (event.target as HTMLElement).focus();
```
Tested this a few more times in FF and Chrome and didn't notice any similar bug to the video above.

**Review by thatblindgeye — COMMENTED** [review]
Just had one followup to one of my other comments above, but otherwise this looks good

**Review by thatblindgeye — APPROVED** [review]

**Review by tlabaj — APPROVED** [review]
